### PR TITLE
updated SI-7331, SI-6843, SI-7731, parser entry point refactoring, assertThrows utility function

### PR DESCRIPTION
**Pull request by densh**
This is a rehash of #2808 that addresses feedback from @gkossakowski and @retronym

### CODE REVIEW

**Comment by adriaanm** [comment]
PLS REBUILD ALL

**Comment by gkossakowski** [comment]
@densh: who is reviewer fro this PR?

**Comment by densh** [comment]
@gkossakowski I guess that @retronym is the reviewer.

File: src/compiler/scala/tools/reflect/quasiquotes/Placeholders.scala [hunk 1]
```diff
@@ -17,18 +17,31 @@ trait Placeholders { self: Quasiquotes =>
   // Step 1: Transform Scala source with holes into vanilla Scala source
 
   lazy val holeMap = new HoleMap()
+  lazy val posMap = mutable.ListMap[Position, (Int, Int)]()
```
Comment (retronym) [inline src/compiler/scala/tools/reflect/quasiquotes/Placeholders.scala:20]: What is is the lifecycle of this map?

Comment (densh) [inline src/compiler/scala/tools/reflect/quasiquotes/Placeholders.scala:20]: It gets initialized during code generation and lays dormant until syntax error happens. In that case it's used to determine equivalent position between quasiquote snippet and code that is fed to the quasiquote parser.

Comment (retronym) [inline src/compiler/scala/tools/reflect/quasiquotes/Placeholders.scala:20]: So basically for the lifetime of a single call to one of the quasiquote macros. I was just checking that this wasn't mixed in to Global or something similar.

```
make(  QuasiquoteClass_api_apply) { case _                                          => _.expandQuasiquote },
    make(QuasiquoteClass_api_unapply) { case _                                          => _.expandQuasiquote }
```

Comment (densh) [inline src/compiler/scala/tools/reflect/quasiquotes/Placeholders.scala:20]: Yes, exactly. It definitely doesn't get mixed into global.

File: test/junit/scala/tools/nsc/symtab/CannotHaveAttrsTest.scala [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+class CannotHaveAttrsTest {
+  class CustomSymbolTable extends SymbolTableForUnitTesting {
+    object CHA extends CannotHaveAttrs {
+      def canEqual(that: Any): Boolean = ???
+      def productArity: Int = ???
+      def productElement(n: Int): Any = ???
+    }
+    val attrlessTrees = List(CHA, EmptyTree, emptyValDef, pendingSuperCall)
+  }
+  def withCtx(body: CustomSymbolTable => Unit) = body(new CustomSymbolTable)
```
Comment (retronym) [inline test/junit/scala/tools/nsc/symtab/CannotHaveAttrsTest.scala]: JUnit creates a new instance of the test for each `@Test` annotated method. (http://martinfowler.com/bliki/JunitNewInstance.html) So you can simple make `CustomSymbolTable` an object and lose the `withCtx` stuff. /cc @gkossakowski

File: test/junit/scala/tools/testing/AssertUtil.scala [hunk 3]
```diff
@@ -0,0 +1,19 @@
+package scala.tools
+package testing
+
+/** This module contains additional higher-level assert statements
+ *  that are ultimately based on junit.Assert primitives.
+ */
+object AssertUtil {
+  /** Check if exception T was thrawn during evaluation of f. If any
```
Comment (retronym) [inline test/junit/scala/tools/testing/AssertUtil.scala]: "thrown"

Comment (retronym) [inline test/junit/scala/tools/testing/AssertUtil.scala]: "Check if exception T (or a subclass)"

File: test/junit/scala/tools/nsc/symtab/CannotHaveAttrsTest.scala (hunk 2 above)
Comment (densh) [inline test/junit/scala/tools/nsc/symtab/CannotHaveAttrsTest.scala]: Good to know, I'll update the tests now.

**Comment by gkossakowski** [comment]
I didn't realize JUnit is smart about class's life cycle. Thanks for the tip @retronym.

**Comment by som-snytt** [comment]
https://github.com/junit-team/junit/wiki/Test-fixtures

I recall (years ago anyway), for N tests, your N instances are created simultaneously, whereas setUp means before test run.

**Comment by retronym** [comment]
LGTM. Great work.

**Pull request merged**